### PR TITLE
asserts etc

### PR DESCRIPTION
**Pull request by eraly**
Biggest change is adding in the neural net as an abstract method to the persistor class and making some assertions: 1) train net with output before freezing 2) output before freezing to output after freezing.

Main motivation was to be able to support nets where we do some training before freezing like the ae example. 

Other minor stuff: Random bug fixes

### CODE REVIEW

File: tests/mlp/test_ae_00.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
+import numpy as np
+import math
+import tensorflow as tf
+from tfoptests.persistor import TensorFlowPersistor
+
+
+class SimpleAE(TensorFlowPersistor):
+    def __init__(self, save_dir, seed=None, base_dir=None, verbose=True):
```
Comment (maxpumperla) [inline tests/mlp/test_ae_00.py]: consider using 

```
    def __init__(self, *args, **kwargs):
        super(SimpleAE, self).__init__(*args, **kwargs)
```

for brevity.

File: tfoptests/persistor.py [hunk 2]
```diff
@@ -40,6 +47,11 @@ def _get_input_shape(self, name):
         '''Get input tensor shape for given node name'''
         raise NotImplementedError
 
+    @abstractmethod
+    def _neural_net(self):
+        '''Implement the neural net and if training return the TF sess. TODO: Document better with example'''
```
[the remark this answers is not in the document]
Comment (maxpumperla) [reply tfoptests/persistor.py]:
```
tfp = TFP(...)

tfp.set_placeholders([in_node])
tfp.set_output_tensors([out_node])
tfp.set_session(sess)  # instead of returning in _neural_net
tfp.set_outputs(...)

tfp.build_and_save_graph()
```

[the remark this answers is not in the document]
Comment (maxpumperla) [reply tfoptests/persistor.py]: if we make those setters return `self`, we can even employ our beloved builder pattern:

```
tfp = TFP(..., train=True) \
  .set_placeholders([in_node])
  .set_output_tensors([out_node])
  .set_session(sess)
  .set_outputs(...)
  .build_and_save_graph()
```

**Comment by eraly** [comment]
Great feedback. Thanks Max. I added in a test_graph class that all the tests should inherit from. And everything that was previously in neural_net is now in the actual test to give visibility into what goes on. Also had the setters return self so it really does look like a builder now. Need to do the docs but let's settle on a framework before I do that.

File: tfoptests/test_graph.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        return None
+
+    ## @abstractmethod
+    def _get_placeholder_shape(self, name):
+        '''Get input tensor shape for given node name'''
+        # raise NotImplementedError
+        return None
+
+    ## @abstractmethod
+    def list_inputs(self):
```
Comment (maxpumperla) [inline tfoptests/test_graph.py]: is this ever not `["input"]`? Maybe it's a good default at least.

File: tfoptests/persistor.py [hunk 4]
```diff
@@ -1,75 +1,56 @@
...
     pass
+import sys
 import os
 import tensorflow as tf
 import numpy as np
 from tensorflow.python.tools import freeze_graph
 
-BASE_DIR = os.environ['DL4J_TEST_RESOURCES'] + '/src/main/resources/tf_graphs/examples'
+# BASE_DIR = os.environ['DL4J_TEST_RESOURCES'] + '/src/main/resources/tf_graphs/examples'
+BASE_DIR = '/Users/susaneraly/SKYMIND/dl4j-test-resources/src/main/resources/tf_graphs/examples'
```
Comment (maxpumperla) [inline tfoptests/persistor.py]: haha, would you mind just exporting that to your path? :)

File: tfoptests/test_graph.py (hunk 3 above)
Comment (eraly) [reply tfoptests/test_graph.py]: Added as a default. We have a few graphs like deep_mnist (input and dropout value) and some of the math_ops where it is not "input". I imagine we will want the flexibility for later as well.

File: tfoptests/persistor.py (hunk 4 above)
Comment (eraly) [reply tfoptests/persistor.py]: whoops. Fixed.
